### ISSUE TITLE
How to use the snackbar action? 

### ISSUE THREAD

**Issue by stefanaerts** [angular/material2#1581]
#### What are the steps to reproduce?

Providing a Plunker (or similar) is the *best* way to get the team to see your issue.
Plunker template: http://plnkr.co/edit/o077B6uEiiIgkC0S06dd

#### Is there anything else we should know?
I wonder where i can find an example of how to use the snackbar. I see the snackbar on the screen,but i have no idea how i can bind an action to the action button on the snackbar.
All i found is this:
  failedAttempt() {
   let config = new MdSnackBarConfig(this.viewContainerRef);
   this.snackBar.open('It didn\'t quite work!', 'Try Again', config);

So how do i bind the Try again in this example to call another method?
Thanks

**Comment by josephperrott** [comment]
If you are using the `.open()` method:

When you create a snackbar using `.open()` you are returned a `MdSnackBarRef` object.
You can watch the observable returned by the `afterDismissed()` method from this ref for when the button was clicked, removing the snack bar. 

Example: http://plnkr.co/edit/ouCdqAkp5dc3veD802lT?p=preview


If you are using the `. openFromComponent()` method:

You could use the same concept as above, but you could also directly call to take any action from within the custom component you have opened.

**Comment by wulfsberg** [comment]
This seems like a dangerous example, since the observable fires on dismiss, not on actual click.
Notably, clicking the "Show snackbar" button while the bar is already on screen causes it to fire, incorrectly stating that "Button was pressed".

It seems like there should be a way to differentiate between the button and, say, a swipe dismiss. (Without having to use a custom component, since the standard API offers a button)

**Comment by stefanaerts** [comment]
Indeed,
In the mean time i found the same problem.
Indeed it just closes 1 time the snackbar and then it stays on status dismissed.
I can't use it to have an action when and only when the user clicks on
the button 'try again' in example.
Because when i go back to the screen the action will happen
automatically again because the dismisses status still applies.
Even when i give the references null, it stays on status dismissed and
the observable will fire all the time the action behind the try again
button.

**Comment by stefanaerts** [comment]
@wulfsberg 
What is the best way at the moment to show a overlay popup at the bottom of my visible part of the screen with an action button? 
See here what i want in this screenshot:
![image](https://cloud.githubusercontent.com/assets/3684599/19686331/e4be3c32-9ac0-11e6-91c7-c3c26eef17bf.png)

**Comment by stefanaerts** [comment]
It should be very easy straight forward to trigger a method when i click on the 'Go to Order' button in the toast. 
Is there nothing available for this?
I guess i have to find it in here .
https://github.com/shlomiassaf/angular2-modal
Nice plunker example,but difficult to understand quickly that custom modal sample.
I wil try to figure it out.
I just need something like in the plunker example of modal
 the method in custom-modal-sample.ts
 beforeClose(): boolean {
    return this.wrongAnswer;
  }
but then 
beforeClose(): void {
goToOrderSummary();
}
I can't find any other overlay( dialog or toast) example with an action possible when you click on the added button.
What is everyone using then?
i looked at ng2toastr, same thing, no action to bind on button
materializecss same, no action to bind on click ok.

**Comment by stefanaerts** [comment]
Thanks Joey for the quick response,
this is what i needed.
It works fine in my app.

Kind rg,
Stefan.

**Comment by stefanaerts** [comment]
In the mean time i found the same problem.
Indeed it just closes 1 time the snackbar and then it stays on status dismissed.
I can't use it to have an action when and only when the user clicks on
the button 'try again in example'.
Because when i go back to the screen the action will happen
automatically again because the dismisses status still applies.
Even when i give the references null, it stays on status dismissed and
the observable will fire all the time the action behind the try again
button.

**Comment by feloy** [comment]
It seems to me that an onAction() observable is missing from the MdSnackBarRef component.

This observable is specified in https://docs.google.com/document/d/1XuD-sfGjZRB-TVFMULmD_FOSsOVJACVRy3XRJsQyYeQ/preview (referenced from #115) but is not yet implemented.

@josephperrott you cannot rely on afterDismissed() observable for this because this observable is called in other situations, at least when a first snackbar is automatically dismissed by a newer snackbar.

**Comment by stefanaerts** [comment]
@josephperrott @wulfsberg @feloy 
I tried it with openFromComponent.
But then i get this error:
A ComponentPortal must have an origin set when attached to a DomPortalHost because the DOM element is not part of the Angular application context

here is the parent component:
![image](https://cloud.githubusercontent.com/assets/3684599/19723636/73ed3f46-9b7c-11e6-92e8-b44e42ece28f.png)

and this is the  dialog example with a yes and no confirm button that i want to open as a snackbar.

![image](https://cloud.githubusercontent.com/assets/3684599/19723736/e47cbcbe-9b7c-11e6-9b6b-4c1652eb6cfb.png)


I get an error on the openFromComponent 
![image](https://cloud.githubusercontent.com/assets/3684599/19723777/1cd8c594-9b7d-11e6-8e15-29e028e37e8f.png)

A ComponentPortal must have an origin set when attached to a DomPortalHost because the DOM element is not part of the Angular application context.


So snackbar doesnt work with a component, dialog(the pizza example) does work,but i want something that overlays at the bottom of my 'VISIBLE' screen, not at the end of the screen after scrolling. Is the snackbar going to work soon,there is a afterClosed() method on md dialog, not on md snackbar. 

I am new to this viewContainer,Overlay stuf i would really appreciate if somebody could help me out here. I like the toast but i want a custom method(a root navigation) to be called when they click on Go To Order.
![image](https://cloud.githubusercontent.com/assets/3684599/19723929/ea17dffe-9b7d-11e6-9c15-8bc0904591ae.png)


Kind Regards
Stefan

**Comment by stefanaerts** [comment]
@feloy @josephperrott  @jelbourn @wulfsberg 

Here a link to a little screen recording of my current status and question.
 I have a snackbar that contains a child component with an action when you click on the snackbar.So it works almost as i want, i have only a problem with the style of the snackbar,
how can i change it?
https://www.youtube.com/watch?v=NjyyHcCA8aM&feature=youtu.be

Thanks.
Kind Rg,
Stefan.

**Comment by feloy** [comment]
@stefanaerts I'm not sure you are attaching your snackbar to the right parent component.
For my part, I'm used to attach it to the top app component. You can get an exemple here: https://github.com/actimeo/simmage-ui/blob/master/src/app/app.component.ts

**Comment by jelbourn** [comment]
Tracking the lack of `onAction` in #1616

Closing this issue since it's asking a question, not reporting a bug/feature

**Comment by stefanaerts** [comment]
@jelbourn, so where should i ask a question and find information related to how to change css of the snackbar?

**Comment by stefanaerts** [comment]
@feloy I moved it to the top appcomponent but that was not the reason the snackbar had the box-sizing: inherit.
I found the reason, i was also experimenting with materializecss and MaterializeModule and included an import in my styles.css, i removed that import and the import of MaterializeModule  and MaterializeDirective and now the snackbar is shown correctly,no box-sizing: inherit any more. Thanks for your time feloy.
